src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,81 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+
+    private List<Product> products;
+    private long maxId;
+
+    public ArrayListProductDao() {
+        this.products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public synchronized Product getProduct(Long id) {
+        return products.stream()
+                .filter(product -> id.equals(product.getId()))
+                .findAny()
+                .orElse(null);
     }
 
     @Override
-    public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+    public synchronized List<Product> findProducts() {
+        return products.stream()
+                .filter(product -> product.getPrice() != null)
+                .filter(product -> product.getStock() > 0)
+                .collect(Collectors.toList());
     }
 
     @Override
-    public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+    public synchronized void save(Product product) {
+        if (product.getId() != null) {
+            products = products.stream()
+                    .peek(p -> {
```

Comment:
Please simplify this statement. Avoid using Stream.peek() method whenever possible. Also, this method just silently does nothing when executed with a Product parameter having non-null id, that is not present in the list